Add usb device ehci controller driver

MarkWangChinese opened the pull request:
NXP EHCI USB controller is used in Kinetis series and RT series for High Speed USB2.0

MaureenHelm commented:
recheck

aurel32 commented:
Thanks @MarkWangChinese . I haven't got time to look at the code yet, but I gave a try to this PR on a MIMXRT1020 board (it only requires a small change to the DTS and the Kconfig file). It works fine for the few tests I have been able to do. However I noticed that the device connects by default in HS mode. USB HS support is not really good in Zephyr yet (see PR #11000), and some host controllers choke on the 64-byte bulk endpoints in HS mode, while HID and INT endpoints usually work fine. You might want to change the default to FS, or even better provide a mechanism in the device-tree to select the maximum speed. You might search for `DT_USBHS_MAXIMUM_SPEED` in `usb_dc_sam.c` or `usb_dc_stm32.c` as an example about how to do that.

MarkWangChinese commented:
> Thanks @MarkWangChinese . I haven't got time to look at the code yet, but I gave a try to this PR on a MIMXRT1020 board (it only requires a small change to the DTS and the Kconfig file). It works fine for the few tests I have been able to do. However I noticed that the device connects by default in HS mode. USB HS support is not really good in Zephyr yet (see PR #11000), and some host controllers choke on the 64-byte bulk endpoints in HS mode, while HID and INT endpoints usually work fine. You might want to change the default to FS, or even better provide a mechanism in the device-tree to select the maximum speed. You might search for `DT_USBHS_MAXIMUM_SPEED` in `usb_dc_sam.c` or `usb_dc_stm32.c` as an example about how to do that.

Thanks @aurel32 , I have added the function to force the EHCI controller to Full-Speed according to usb_dc_sam.c or usb_dc_stm32.c

MaureenHelm commented:
@finikorg 	@jfischer-phytec-iot 	please review

MaureenHelm commented on drivers/usb/device/usb_dc_mcux_ehci.c:
```diff
@@ -0,0 +1,490 @@
+#include <device.h>
+#include "usb_dc_mcux_nxp.h"
+#include "usb_device_ehci.h"
+#include <logging/sys_log.h>
```
Need to switch to new logging system. This will become:
```
#include <logging/log.h>
LOG_MODULE_REGISTER(usb_dc_mcux_ehci, CONFIG_USB_DRIVER_LOG_LEVEL);
```

MarkWangChinese replied:
@MaureenHelm thanks, I have updated it.

MaureenHelm commented on drivers/usb/device/usb_dc_mcux_ehci.c:
```diff
@@ -0,0 +1,490 @@
+		    usb_isr_handler, 0, 0);
+	irq_enable(DT_USBD_MCUX_EHCI_IRQ);
+	s_Device.attached = 1;
+	SYS_LOG_DBG("attached");
```
This should be `LOG_DBG`

MarkWangChinese replied:
@MaureenHelm thanks, I have updated it.

MaureenHelm commented on drivers/usb/device/usb_dc_mcux_nxp.h:
```diff
@@ -0,0 +1,372 @@
+    #define USB_DEVICE_CONFIG_LPCIP3511FS (0U)
+#endif
+
+/* IP3511 HS instance count */
```
Why do we need any of these instance count macros?

MarkWangChinese replied:
@MaureenHelm For example: USB_DEVICE_CONFIG_EHCI, I convert CONFIG_USB_DC_NXP_EHCI to USB_DEVICE_CONFIG_EHCI in the head file.
In MCUXpresso SDK USB controller driver, USB_DEVICE_CONFIG_EHCI is used to define the EHCI controller driver instance count (if it is 0, the function is disable), In Zephyr USB device stack, the multiple instances is not supported yet, CONFIG_USB_DC_NXP_EHCI only means enable or disable. In order to support multiple instance in furture and do less changes of original MCUX USB device controller driver (usb_device_ehci.c), It may be better to keep this convertion.

MaureenHelm replied:
Ok, this makes more sense after you moved `usb_dc_mcux_nxp.h` to `ext/`

MaureenHelm commented on drivers/usb/device/usb_dc_mcux_nxp.h:
```diff
@@ -0,0 +1,372 @@
+
+#if defined(CONFIG_BIG_ENDIAN)
+#define SWAP2BYTE_CONST(n) ((((n) & 0x00FFU) << 8U) | (((n) & 0xFF00U) >> 8U))
+#define SWAP4BYTE_CONST(n) ((((n) & 0x000000FFU) << 24U) | (((n) & 0x0000FF00U) << 8U) | (((n) & 0x00FF0000U) >> 8U) | (((n) & 0xFF000000U) >> 24U))
```
Use the macros in `include/misc/byteorder.h` instead

MarkWangChinese replied:
@MaureenHelm thanks, I have updated it.

MaureenHelm commented on drivers/usb/device/usb_dc_mcux_nxp.h:
```diff
@@ -0,0 +1,372 @@
+/* USB controller ID */
+typedef enum _usb_controller_index {
+	/* KHCI 0U */
+	kUSB_ControllerKhci0            = 0U,
```
These look like MCUXpresso SDK enums. Should these defined somewhere in `ext/hal/nxp/mcux`?

MarkWangChinese replied:
@MaureenHelm thanks, I have updated it.

MaureenHelm commented on soc/arm/nxp_imx/rt/Kconfig.defconfig.mimxrt1052:
```diff
@@ -44,6 +44,13 @@ config I2C_MCUX_LPI2C
+if USB
+
+config USB_DC_NXP_EHCI
+	default y
```
This can move up a level to `soc/arm/nxp_imx/rt/Kconfig.defconfig.series`. See commit 501db1e0b96133cd5377c518e681e8f82114f836

MarkWangChinese replied:
@MaureenHelm thanks, I have updated it.

MaureenHelm commented:
Please add usb_device to the board supported list, like this:
https://github.com/zephyrproject-rtos/zephyr/blob/732cfac519ee7822b0d19eba5ef946f30649204d/boards/arm/frdm_k64f/frdm_k64f.yaml#L9-L16

This will make CI build the new usb driver.

agansari reviewed (changes requested):
Driver layer looks okay, there are some other minor changes required.
I've tried testing it on my 1050 but it did not manage to get it working, @aurel32 what changes have you made?
Need any help with this pull request?

agansari commented on soc/arm/nxp_imx/rt/soc.c:
```diff
@@ -141,6 +151,17 @@ static ALWAYS_INLINE void clkInit(void)
+				CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
+	CLOCK_EnableUsbhs0Clock(kCLOCK_Usb480M,
+				CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC);
+	usb_phy_config_struct_t phyConfig = {
```
better rename to usbPhyConfig so it's not to be confused with ethernet phy.

MarkWangChinese replied:
@agansari  thanks, I have updated it.

aurel32 commented:
> I've tried testing it on my 1050 but it did not manage to get it working, @aurel32 what changes have you made?

That's the patch I used. I haven't tried it recently though, and when I tried the default RAM was still DTCM0.

```
diff --git a/boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts b/boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts
index 2faf1c543d..2b17b04b04 100644
--- a/boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts
+++ b/boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts
@@ -85,3 +85,7 @@ arduino_serial: &uart2 {};
                status = "ok";
        };
 };
+
+&usbd1 {
+       status = "ok";
+};
diff --git a/soc/arm/nxp_imx/rt/Kconfig.defconfig.mimxrt1021 b/soc/arm/nxp_imx/rt/Kconfig.defconfig.mimxrt1021
index ec1e6b41c6..c4a369cf46 100644
--- a/soc/arm/nxp_imx/rt/Kconfig.defconfig.mimxrt1021
+++ b/soc/arm/nxp_imx/rt/Kconfig.defconfig.mimxrt1021
@@ -23,4 +23,11 @@ config IPG_DIV
 config GPIO
        default y
 
+if USB
+
+config USB_DC_NXP_EHCI
+       default y
+
+endif # USB
+
 endif # SOC_MIMXRT1021

```

MarkWangChinese commented on drivers/usb/device/Kconfig:
```diff
@@ -72,6 +72,13 @@ config USB_KINETIS
 
+config USB_DC_NXP_EHCI
+	bool "Kinetis and RT EHCI USB Device Controller Driver"
+	depends on SOC_FAMILY_IMX
```
@MaureenHelm thanks, I have updated it.

MarkWangChinese commented on drivers/usb/device/CMakeLists.txt:
```diff
@@ -4,3 +4,6 @@ zephyr_sources_ifdef(CONFIG_USB_DC_SAM0  usb_dc_sam0.c)
 zephyr_sources_ifdef(CONFIG_USB_NRF52840 usb_dc_nrfx.c)
 zephyr_sources_ifdef(CONFIG_USB_KINETIS  usb_dc_kinetis.c)
+zephyr_sources_ifdef(CONFIG_USB_DC_NXP_EHCI  usb_dc_mcux_ehci.c)
+zephyr_include_directories(./)
```
Thanks, I have updated it.

MarkWangChinese commented on soc/arm/nxp_imx/rt/soc.c:
```diff
@@ -141,6 +151,17 @@ static ALWAYS_INLINE void clkInit(void)
+	usb_phy_config_struct_t phyConfig = {
+		BOARD_USB_PHY_D_CAL, BOARD_USB_PHY_TXCAL45DP, BOARD_USB_PHY_TXCAL45DM,
+	};
+	USB_EhciPhyInit(2, 24000000U, &phyConfig);
```
@agansari thanks, I have updated it.

MarkWangChinese commented:
> Driver layer looks okay, there are some other minor changes required.
> I've tried testing it on my 1050 but it did not manage to get it working, @aurel32 what changes have you made?
> Need any help with this pull request?

@agansari RT1050 changed the flash and ram as external flash and sdram, the original patch run fail, I have updated the patch and run the hid-mouse example successfully on RT1050.

MarkWangChinese commented on soc/arm/nxp_imx/rt/arm_mpu_regions.c:
```diff
@@ -20,7 +20,14 @@ static const struct arm_mpu_region mpu_regions[] = {
+				(NORMAL_OUTER_INNER_NON_CACHEABLE_NON_SHAREABLE | \
+				 MPU_RASR_XN_Msk | REGION_SRAM_0_SIZE | P_RW_U_NA_Msk) \
+			 }),
+#else
```
@MaureenHelm USB examples need disable the cacheable of RAM. I cannot find the better way to change the RAM cacheable attribute, I do as the above changes. could you please help review the changes?

aurel32 replied:
@MarkWangChinese You probably need to enable NOCACHE support and define the area you want to be non-cacheable with __nocache. Look for example what is done in the Ethernet MCUX driver.

agansari replied:
@MarkWangChinese you need the following to use cache and nocache in your usb driver: 
1. add USB_DEVICE_CONFIG_BUFFER_PROPERTY_CACHEABLE just like FSL_SDK_ENABLE_DRIVER_CACHE_CONTROL
2. select NOCACHE_MEMORY and ARM_MPU when configuring USB_DC_NXP_EHCI, just like ETH_MCUX's menuconfig
3. #define USB_CONTROLLER_DATA __nocache similar to SDK examples
4. ep_buf_pool set as __nocache: __nocache K_MEM_POOL_DEFINE(ep_buf_pool, 16, 1024, EP_BUF_NUMOF_BLOCKS, 4); this looks strange but the definition works.

MarkWangChinese replied:
@agansari @aurel32  Thanks, I have updated it, you can review it.

agansari reviewed (approved):
Code looks good, implementation was validated on my setup.
I have a few minor observations, also not sure about multiple:
`zephyr_include_directories(.)`

MarkWangChinese commented on ext/hal/nxp/mcux/middleware/usb/device/port/zephyr/usb_dc_mcux.h:
```diff
@@ -0,0 +1,366 @@
+/******************************************************************************
+ * Definitions
+ *****************************************************************************/
+/* KHCI instance count */
```
yes, it is shared with KHCI and LPC IP3511, I will add KHCI and LPC IP3511 controller driver support after this PR done.

MarkWangChinese commented:
> Please add usb_device to the board supported list, like this:
> [zephyr/boards/arm/frdm_k64f/frdm_k64f.yaml](https://github.com/zephyrproject-rtos/zephyr/blob/732cfac519ee7822b0d19eba5ef946f30649204d/boards/arm/frdm_k64f/frdm_k64f.yaml#L9-L16)
> 
> Lines 9 to 16 in [732cfac](/zephyrproject-rtos/zephyr/commit/732cfac519ee7822b0d19eba5ef946f30649204d)
> 
>  supported: 
>    - netif:eth 
>    - adc 
>    - i2c 
>    - nvs 
>    - spi 
>    - gpio 
>    - usb_device 
> This will make CI build the new usb driver.

I have added usb_device to rt1050 board.

MarkWangChinese commented:
Hi @MaureenHelm what's the status of this pull request? Do I need do more changes or wait the approvement? Thanks

MaureenHelm commented on ext/hal/nxp/mcux/middleware/usb/device/port/zephyr/usb_dc_mcux.h:
```diff
@@ -0,0 +1,370 @@
+#define USB_LONG_TO_LITTLE_ENDIAN(n) (n)
+#define USB_SHORT_FROM_LITTLE_ENDIAN(n) (n)
+#define USB_LONG_FROM_LITTLE_ENDIAN(n) (n)
+#endif
```
You can replace lines 65-75 with:
```
#define USB_SHORT_TO_LITTLE_ENDIAN(n)	sys_cpu_to_le16(val)
#define USB_LONG_TO_LITTLE_ENDIAN(n)	sys_cpu_to_le32(val)
#define USB_SHORT_FROM_LITTLE_ENDIAN(n)	sys_le16_to_cpu(val)
#define USB_LONG_FROM_LITTLE_ENDIAN(n)	sys_le32_to_cpu(val)
```

MaureenHelm commented on ext/hal/nxp/mcux/middleware/usb/device/port/zephyr/usb_dc_mcux.h:
```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright (c) 2018 NXP Corporation
```
Remove `Corporation`

MarkWangChinese replied:
Removed

MaureenHelm reviewed (commented):
It's not quite clear if `usb_dc_mcux.h` was imported from the MCUXpresso SDK. The coding style suggests it was imported, but there are some zephyr-specific lines too. Can you split it up into two files? Keep the SDK license (BSD-3) on the imported files, and use the Apache license on new files.

> Hi @MaureenHelm what's the status of this pull request? Do I need do more changes or wait the approvement? Thanks

I'd like to get @jfischer-phytec-iot or @finikorg to review before we consider merging this. If they approve, then we will need to bring this to the TSC for a vote because the 1.14 merge window is already closed to new features.

MarkWangChinese commented:
> It's not quite clear if `usb_dc_mcux.h` was imported from the MCUXpresso SDK. The coding style suggests it was imported, but there are some zephyr-specific lines too. Can you split it up into two files? Keep the SDK license (BSD-3) on the imported files, and use the Apache license on new files.
> 
> > Hi @MaureenHelm what's the status of this pull request? Do I need do more changes or wait the approvement? Thanks

I have changed license of usb_dc_mcux.h as MCUXpresso SDK license. Most of the content are from MCUXpresso SDK and few are modified based on MCUXpresso SDK, only the follow lines use the zephyr's kconfig MACROs.
#ifdef CONFIG_USB_DC_NXP_EHCI
    #define USB_DEVICE_CONFIG_EHCI (1U)
#else
    #define USB_DEVICE_CONFIG_EHCI (0U)
#endif

MarkWangChinese commented:
Hi @MaureenHelm, is the pull request blocked to wait the 1.14 finish? Do I need do more changes or just  wait? Thanks

MaureenHelm reviewed (changes requested):
> Hi @MaureenHelm, is the pull request blocked to wait the 1.14 finish? Do I need do more changes or just wait? Thanks

Hi @MarkWangChinese, the 1.14 release is finished and the merge window is now open again.

MarkWangChinese committed:
ext: mcux: Import NXP SDK USB EHCI controller driver to ext/ directory.

Put the driver files to ext/hal/nxp/mcux/middleware/usb/device.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
drivers: usb: nxp: add NXP EHCI USB device controller driver shim

usb_dc_mcux_ehci driver is one shim of the NXP SDK ehci driver.
select NOCACHE_MEMORY if HAS_MCUX_CACHE

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
soc: nxp_imx: configure USB device EHCI controller driver for rt1050

add usbd1 definition to rt dts file,
set EHCI controller config default value in rt1050 default config file,
add EHCI controller driver MACROs to dts_fixup.h,
initialize EHCI clock in rt soc.c
add HAS_MCUX_USB_EHCI for supported soc in Kconfig.soc

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

MarkWangChinese committed:
boards: mimxrt1050_evk: configure USB device EHCI driver for rt1050

set usbd1 status as OK in mimxrt1050_evk.dts
add usb_device supported information in rt1050 yaml file.

Signed-off-by: Mark Wang <yichang.wang@nxp.com>

carlescufi commented:
@MarkWangChinese can you address the review comments?

MarkWangChinese commented on boards/arm/mimxrt1050_evk/mimxrt1050_evk.yaml:
```diff
@@ -20,3 +20,4 @@ supported:
   - spi
   - netif:eth
   - hwinfo
+  - usb_device
```
Hi @MaureenHelm , do you mean the frdmk64? I don't get it and I don't know how to add it.

MarkWangChinese replied:
This pull request are only for the USB EHCI controller. RT1050 and RT1060 have this controller, Frdmk64 doesn't have it, the pull-request is not related to the frdmk64.

MaureenHelm replied:
Sorry, my mistake. I meant `boards/arm/mimxrt1050_evk/doc/index.rst`

MarkWangChinese commented on ext/hal/nxp/mcux/README:
```diff
@@ -84,3 +84,5 @@ Patch List:
    * Automatically enable ENET_ENHANCEDBUFFERDESCRIPTOR_MODE in HAL if gPTP
      support is enabled.
+
+   * Add USB device controller drivers.
```
Hi @MaureenHelm , how to add it? I just use the SDK USB stack common files, they don't belong to any boards.

MaureenHelm replied:
Pick the latest SDK release tag. We already have references to `REL_2.5.0_RT1015_RFP` and `REL_2.5.0_REL9_RFP_RC3_7_1` in `ext/hal/nxp/mcux/README`; can you use one of those?

MaureenHelm replied:
I still don't see the tag listed in the README or commit message. Will you please add it?

MarkWangChinese commented on drivers/usb/device/Kconfig:
```diff
@@ -80,6 +80,15 @@ config USB_KINETIS
+	depends on HAS_MCUX_USB_EHCI
+	select USB_DEVICE_DRIVER
+	select NOCACHE_MEMORY if HAS_MCUX_CACHE
+	select ARM_MPU if CPU_CORTEX_M7
```
yes, I will do these

MarkWangChinese commented on soc/arm/nxp_imx/rt/Kconfig.soc:
```diff
@@ -25,6 +25,7 @@ config SOC_MIMXRT1021
 	select INIT_SYS_PLL
 	select INIT_USB1_PLL
 	select INIT_ENET_PLL
+	select HAS_MCUX_USB_EHCI
```
OK

MaureenHelm commented on drivers/usb/device/Kconfig:
```diff
@@ -82,6 +82,14 @@ config USB_KINETIS
+	bool "Kinetis and RT EHCI USB Device Controller Driver"
+	depends on HAS_MCUX_USB_EHCI
+	select USB_DEVICE_DRIVER
+	select NOCACHE_MEMORY if HAS_MCUX_CACHE
```
Please update the commit message to reflect that you're no longer adding `select ARM_MPU if CPU_CORTEX_M7`.

MaureenHelm reviewed (changes requested):
Please fix the shippable and kconfig failures, don't worry about the license failure.

raveslave commented on dts/arm/nxp/nxp_rt.dtsi:
```diff
@@ -327,6 +327,17 @@
+			interrupts = <113 1>;
+			interrupt-names = "usb_otg";
+			num-bidir-endpoints = <8>;
+			maximum-speed = "full-speed";
```
shouldn't this be high-speed?

aurel32 replied:
The Zephyr USB stack is not fully HS compatible, for example bulk endpoints do not have the correct size in HS speed. It's therefore a good idea to limit the speed to FS by default, an app fully validated on HS can override the default.
